File: src/parser.nim
import ast
import lexing

import strformat, tables, macros, options, sugar, sets, sequtils

template err(msg: string) = quit fmt"{getStackTrace()}{self.pos.line}:{self.pos.col}: " & msg  # TODO

type
  # All of this is to be handled exclusively from the `parseExpr` function.
  # Individual parsing functions then check `.exprStyle` via `thisExprCan`.
  ExprCan* = enum
    canColon,        # Can use `:` calls
    canAssg,         # Can do assignments
    canBlock,        # Can have an implicit block *at all*
    canComma,        # Can use un-parenthesized function calls of multiple args
    canImplicitBlock,# Can start with an Indent (i.e the entire expr is one implicit block)
  ExprStyle* = set[ExprCan]

const
  NormalExpr*: ExprStyle = { canColon, canBlock, canComma }
  AssingableExpr*: ExprStyle = NormalExpr + { canAssg }
  UnblockedExpr* = NormalExpr - { canBlock }
  # Anywhere you can implicit block, you could also pass an assignment.
  MaybeBlockExpr* = NormalExpr + { canAssg, canImplicitBlock }

type Parser* = ref object
  exprStyle*: ExprStyle
  lexer*: Lexer

  cur*: Token
  # Position of cur
  pos*: Pos
  # Reference position. Use it to figure out if we in/dedented or newlined.
  # Must be manually set.
  # Must be cleared by one of:
  #   - Compound parser
  #   - Block parser
  refPos*: tuple[line: int, level: int]
  # Was cur directly (no whitespace) adjacent to the previous token?
  attached*: bool
  prec*: int
  isControl*: bool
  isBin*: bool
  isAssg*: bool
  isAccess*: bool

  # Do we currently care about newlines/indents/dedents?
  blocking*: bool
  # Did we parse something like a control statement which unambiguously
  # ends the current expression?
  stopped*: bool

macro makePrecs*(body: untyped): untyped =
  result = newCall(ident "toTable", newTree(nnkTableConstr, []))

  var i = 0
  for level in body:
    for op in level:
      result[1].add newColonExpr(op, newIntLitNode i)
    inc i

# Binary operator precedence cache
# Words must be manually added, new sigils will be added based on
# what they look like when they're found.
var Precedences* = makePrecs [
  # Level 0 is assignment
  [iLambda],
  [iOr, iXor],
  [iAnd],
  [iEq, iNeq, iGtEq, iLtEq, iLt, iGt],
  [iAdd, iSub],
  [iMul, iDiv, iMod],
]
# Don't treat pure `:` as a binop

proc getMaxPrec*(): int =
  result = 0
  for v in Precedences.values:
    if v > result: result = v
  result.inc
var MaxPrec* = getMaxPrec()


var AssgOps*: HashSet[StrID] = toHashSet [
  iAssg, iAddAssg, iSubAssg, iMulAssg, iDivAssg
]


# Although technically binary, access ops are treated differently
var AccessOps*: HashSet[StrID] = toHashSet [
  iAccess, iAccessPipe, iOptAccess, iOptAccessPipe,
]



# Words that split into their own special handlers
# Each is added directly below its function

type ControlFunc* = (proc(self: Parser): Atom)
var ControlWords* = initTable[StrID, ControlFunc]()
proc registerHandler*(words: openarray[StrID], handler: ControlFunc) =
  for word in words:
    if ControlWords.contains word:
      quit "ERROR: Re-registering control word " & $word
    ControlWords[word] = handler

# -1 is no prec
proc precOf*(t: Token): int =
  result = -1
  case t.kind:
    of tkWord:
      if Precedences.contains t.id:
        result = Precedences[t.id]
    of tkSigil:
      if Precedences.contains t.id:
        result = Precedences[t.id]
      else:
        if ControlWords.contains(t.id) or AccessOps.contains(t.id) or AssgOps.contains(t.id): result = -1
        else:
          # echo fmt"Precedence calculation is TODO ({t})"
          Precedences[t.id] = -1
    else: discard

# I try to keep a record of the syntax for each proc. Some conventions for those:
# - I mostly ignore whitespace in these. The comments are for the big picture
# - I use a mixture of EBNF, regexes, and custom syntax
# - Actual words refer to something else
# - Things in quotes refer to the literal text
# - {} is 0+
# - [] is 0-1
# - ?word predicates an optional [] or {} on word's value
# - ~word means find but do not match word

proc advance*(self: Parser): Token =
  result = self.cur
  # In ligi, could be
  # self.(pos, attached, cur) = self.lexer.scan
  (self.pos, self.attached, self.cur) = self.lexer.scan

  self.isBin = false
  self.isAssg = false
  self.isControl = false
  self.isAccess = false

  if self.cur.kind in { tkWord, tkSigil }:
    self.prec = precOf self.cur
    if self.prec != -1:
      self.isBin = true
    elif ControlWords.contains self.cur.id:
      self.isControl = true
    elif AccessOps.contains self.cur.id:
      self.isAccess = true
    elif AssgOps.contains self.cur.id:
      self.isAssg = true
  # echo "Advanced to ", self.cur, fmt"({self.pos}, {self.refPos})"

# Commands for working with indentation
template newlined*(): bool {.dirty.} =
  self.blocking and self.pos.line != self.refPos.line
#template samelined(): bool {.dirty.} = not newlined 
template indented*(): bool {.dirty.} =
  self.blocking and self.pos.level > self.refPos.level
template dedented*(): bool {.dirty.} =
  self.blocking and self.pos.level < self.refPos.level

proc nextIs*(self: Parser, i: StrID, respectNewline = true): bool =
  result = not (newlined and respectNewline) and self.cur.kind in {tkWord, tkSigil, tkPunc} and self.cur.id == i
proc nextIs*(self: Parser, tags: HashSet[StrID], respectNewline = true): bool =
  result = not (newlined and respectNewline) and self.cur.kind in {tkWord, tkSigil, tkPunc} and self.cur.id in tags
proc nextIs*(self: Parser, ids: openArray[StrID], respectNewline = true): bool =
  if newlined and respectNewline: return false
  result = false
  if self.cur.kind in {tkWord, tkSigil, tkPunc}:
    for i in ids:
      if self.cur.id == i:
        return true
proc nextIs*(self: Parser, kinds: set[TokenKind], respectNewline = true): bool =
  not (newlined and respectNewline) and self.cur.kind in kinds
proc nextIs*(self: Parser, kind: TokenKind): bool = not newlined and self.cur.kind == kind


template nextLineIs*(thing: untyped): bool = self.nextIs(thing, false)
template nextIs*(thing: untyped): bool = self.nextIs(thing)
template take*(thing: untyped): Token =
  if not self.nextIs thing:
    when thing is StrID:
      err "Expected " & thing.lookup
    else:
      err "Expected " & $thing
  self.advance
template match*(thing: untyped) = discard take thing
template tryMatch*(thing: untyped): bool =
  if nextIs thing:
    match thing
    true
  else:
    false


template setDent*() {.dirty.} =
  # echo "SetDent to ", self.pos.level
  self.refPos.level = self.pos.level
template setLine*() {.dirty.} =
  # echo "SetLine to ", self.pos.line
  self.refPos.line = self.pos.line

template pushRef*(body: untyped) =
  let oldRefPos = self.refPos
  body
  self.refPos = oldRefPos
  # echo fmt"Back to {self.refPos}"

template setBlocking(body: untyped, wth: bool): untyped =
  let oldBlocking = self.blocking
  self.blocking = wth
  block:
    body
  self.blocking = oldBlocking
template withBlocking*(body: untyped): untyped = setBlocking(body, true)
template withoutBlocking*(body: untyped): untyped = setBlocking(body, false)

# Simplifies making sure we check stopped each time
template descend*(body: untyped) =
  body
  if self.stopped: return

template thisExprCan*(can: set[ExprCan]): bool = can <= self.exprStyle

# The fundamental funcs
proc parseExpr*(self: Parser, kind: ExprStyle): Atom
proc parseTypeDesc*(self: Parser, trailing: bool): Atom
proc parseBlock*(self: Parser): Atom


template parseDelimited*(delim: StrID, endCond: untyped): seq[Atom] =
  var result: seq[Atom]
  # echo "Parsing a `" & delim.lookup & "` delimited list"
  while not endCond:
    if nextIs delim:
      match iComma
      continue
    # echo fmt"Parsing expr when next is {self.cur}"
    result.add(self.parseExpr(UnblockedExpr))
    self.stopped = false
    if not nextIs delim:
      # echo "Breaking"
      break
    match delim
  result

proc parseTypeDesc*(self: Parser, trailing: bool): Atom =
  if not nextIs iColon:
    return iSink.toAtom

  let innerKind =
    if trailing:
      NormalExpr - { canBlock, canColon }
    else:
      NormalExpr

  match iColon
  result = self.parseExpr innerKind

  if trailing: match iColon

proc parseBlock*(self: Parser): Atom =
  result = list(ibBlock, [])
  withoutBlocking:
    result.add self.parseTypeDesc(true)

  # Note that we setDent *after* the typedesc.
  # Thus you can do `{ :i32:
  #   code...
  # }`, where normally you couldn't put the `:i32:` at a "different" level from the main body.
  # 
  withBlocking: pushRef:
    setDent
    while not (dedented or self.cur.kind == tkEOF):
      setLine
      result.children.add self.parseExpr(NormalExpr)
      self.stopped = false




proc parseAccessible*(self: Parser): Atom =
  if nextIs { tkWord, tkStrop }:
    result = Atom(
      kind: akWord,
      id: self.advance.id,
    )
  elif nextIs { tkStr }:
    result = Atom(
      kind: akStr,
      str: self.advance.str,
    )
  # For the purposes of unaries, treat as words
  elif nextIs { tkSigil }:
    result = Atom(
      kind: akWord,
      id: self.advance.id,
    )
  else:
    if newlined:
      err "Hit a newline. Expected a word or a string."
    else:
      err fmt"Expected a word, sigil, or a string. Found {self.cur}"
  # echo fmt"Got a {result}"

template handlePostOps() =
  # echo fmt"Parsing a postop on {self.cur}"
  if nextIs iLParen:
    withoutBlocking:
      match iLParen
      result = list(
        result,
        parseDelimited(iComma, nextIs iRParen)
      )
      match iRParen
  elif nextIs iLBracket:
    withoutBlocking:
      match iLBracket
      result = list(ibAt, result)
      result.children = result.children.concat parseDelimited(iComma, nextIs iRBracket)
      match iRBracket
  elif self.isAccess:
    result = list(
      self.advance.id, # `.`, `?.`, etc
      result,
      self.parseRedirectable(mayAccess=false)
    )
  else:
    err "Improper usage of handlePostOps"
# Handles all table driven stuff
proc parseRedirectable*(self: Parser, mayAccess = true): Atom =
  # First check needed to stop `"if"` (a string) from being a keyword
  if self.cur.kind in {tkWord, tkPunc} and ControlWords.contains self.cur.id:
    # `if`, `while`, `let`, `{} blocks`, etc
    # All control structures set stopped, but stuff like `{}` or `()` don't.
    descend:
      # echo "Parsing control ", self.cur.id.lookup
      result = ControlWords[self.cur.id](self)
      if self.stopped and not mayAccess:
        err "May not have an unambiguous stop in a non-accessible redirectable."
  else:
    # Default handlers for strings, chars, and non-control words
    result = self.parseAccessible()
  while mayAccess and not newlined and (self.isAccess or (self.attached and nextIs [iLParen, iLBracket])):
    handlePostOps()

  if nextIs(iColon) and thisExprCan {canColon}:
    match iColon
    result = list(
      result,
      self.parseExpr (NormalExpr + {canImplicitBlock})
    )
    self.stopped = true

  elif indented and self.isAccess:
    # echo "In here"
    pushRef:
      setDent
      while not dedented:
        if not self.isAccess:
          err "Expected an access op"
        setLine
        handlePostOps()
    self.stopped = true
  elif nextIs iColon:
    self.stopped = true

proc parseUnary(self: Parser): Atom =
  descend: result = self.parseRedirectable()
  # Parsing things like `if` or `when` will self.stopped and properly prevent
  # us from treating them as unary operators. Thus, we can fearlessly treat anything
  # that gets here as "unary operator"-able
  # 
  # If the next op is a marked bin/assgop, we need to fall back to parseBinary
  if not (self.isBin or self.isAssg or newlined) and self.cur.kind notin { tkPunc }:
    result = list( result, self.parseUnary() )
    if nextIs(iComma) and thisExprCan {canComma}:
      self.exprStyle = self.exprStyle - {canComma}
      while tryMatch iComma:
        result.children.add self.parseUnary()
      self.stopped = true

proc parseBinary(self: Parser, prec = 0): Atom =
  template nextLevel(): Atom =
    if prec == MaxPrec: self.parseUnary
    else: self.parseBinary(prec + 1)
  descend: result = nextLevel()
  while self.cur.kind in {tkWord, tkSigil} and Precedences.getOrDefault(self.cur.id, -1) == prec:
    descend:
      result = list( self.advance.id, result, nextLevel() )

proc parseExpr*(self: Parser, kind: ExprStyle): Atom =
  let oldStyle = self.exprStyle
  self.exprStyle = kind


  if indented and thisExprCan {canImplicitBlock}:
    result = self.parseBlock()
    self.stopped = true
  else:
    result = self.parseBinary()
    if nextIs(AssgOps) and thisExprCan {canAssg}:
      # Cannot chain assignments.
      result = list( self.advance.id, result, self.parseExpr(self.exprStyle - {canAssg}) )

  self.exprStyle = oldStyle


proc parse*(lexer: Lexer): Atom =
  var self = Parser(
    lexer: lexer,
    pos: (-1, -1, 0),
    refPos: (-1, 0),
    attached: false,
    prec: -1,
    isControl: false,
    isBin: false,
    isAccess: false,
    blocking: true,
    stopped: false,
  )
  discard self.advance
  self.parseBlock

File: src/Parser.zig
pub const Parser = @This();

const std = @import("std");
const printf = std.debug.warn;
const Alloc = std.mem.Allocator;
const List = std.ArrayList;

const StrCache = @import("StrCache.zig");
const StrID = StrCache.StrID;

const Lexer = @import("Lexer.zig");
const FilePos = Lexer.FilePos;
const ScanRes = Lexer.ScanRes;
const Token = Lexer.Token;
const CToken = Lexer.CToken;
const ast = @import("ast.zig");
const Atom = ast.Atom;
const common = @import("common_ast.zig");
const CommonID = common.CommonID;
const TokenDict = Lexer.TokenDict;

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║Accompanying Types                                                         ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
pub const ParseHook = fn (*Parser) ParseError!?Atom;

pub const PrecKind = enum { none, binary, access, assg };
pub const CPrec = struct { tok: CToken, prec: PrecKind = .binary };
pub const Prec = union(PrecKind) {
    none: void,
    binary: usize,
    access: void,
    assg: void,

    pub fn eql(p1: Prec, p2: Prec) bool {
        return switch (p1) {
            .binary => p1.binary == p2.binary,
            .access => p2 == .access,
            .assg => p2 == .assg,
        };
    }
};

pub const Matchable = union(enum) {
    token: Token,
    common: union(enum) {
        word: CommonID,
        punc: CommonID,

        pub fn toStr(self: @This()) []const u8 {
            return common.resolve(common.resolveCommon(switch (self) {
                .word => |x| x,
                .punc => |x| x,
            }));
        }
    },
    prec: Prec,
    /// Anything that can be a binary operator
    any_binary: void,
    /// Anything that can be an access operator
    any_access: void,
    any_assg: void,
    any_word: void,

    any_of: []const Matchable,
    one_of: []const Matchable,
    seq_of: []const Matchable,

    pub fn eql(self: Matchable, tok: Token, parser: *const Parser) bool {
        switch (self) {
            .token => |t| return tok.eql(t),
            .common => |com| switch (com) {
                .word => |id| return tok.eql(.{ .word = common.resolveCommon(id) }),
                .punc => |id| return tok.eql(.{ .punc = common.resolveCommon(id) }),
            },
            .prec => |p1| {
                return if (parser.precs.get(tok)) |p2| p1.eql(p2) else false;
            },
            .any_binary => {
                return if (parser.precs.get(tok)) |p| p == .binary else false;
            },
            .any_access => {
                return if (parser.precs.get(tok)) |p| p == .access else false;
            },
            .any_assg => {
                return if (parser.precs.get(tok)) |p| p == .assg else false;
            },
            .any_word => return tok == .word,
            else => return false,
        }
    }
};

pub const ParseError = error{
    Unmatched,
    OutOfMemory,
    Unhandlable,
    RightNotFound,
};

pub const TokInfo = struct {
    tok: Token,
    pos: FilePos,
    attached: bool,
    prec: Prec,

    pub fn from(s: ScanRes, p: *Parser) TokInfo {
        return .{
            .tok = s.tok,
            .pos = s.pos,
            .attached = s.attached,
            .prec = p.precOf(s.tok, s.attached),
        };
    }
};

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║Fields                                                                     ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
cache: *StrCache,
lexer: *Lexer,

toks: List(TokInfo),
/// The level we expect to be at. Will be routinely pushed back by functions, with the new one
/// being popped back on with a `defer`.
ref_level: usize,
/// Set if something unambiguously ended the current expression level.
/// Parsers like tuples/blocks/etc should unset this flag before moving to the
/// next value in the chain.
/// For example, control structures should usually set this, as they have no concrete end token.
ended: bool = false,

/// Parsed just below the unary level. Allows for things like `if`, `{}`, `fn`, `$`, etc.
hooks: TokenDict(ParseHook),
precs: TokenDict(Prec),

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║Interface                                                                  ║
// ╚═══════════════════════════════════════════════════════════════════════════╝

pub fn init(lexer: *Lexer, alloc: *Alloc) Parser {
    return .{
        .ref_level = 0,
        .lexer = lexer,
        .cache = lexer.cache,
        .toks = List(TokInfo).init(alloc),
        .hooks = TokenDict(ParseHook).init(alloc),
        .precs = TokenDict(Prec).init(alloc),
    };
}
pub fn advanceN(self: *Parser, n: usize) ParseError!void {
    var i: usize = 0;
    while (i < n) _ = try self.advance();
}

/// Peek at the nth token without actually consuming anything.
pub fn nth(self: *Parser, n: usize) !?TokInfo {
    var res: ?Token = null;
    while (n >= self.cur.items.len) {
        const next = try self.lexer.scan();
        if (next) |item| {
            try self.cur.items.append(item);
        } else break;
    }
    if (self.cur.items.len > n) {
        res = self.cur.items[n];
    }

    return res;
}

/// Actual advancing will happen when we next call nth(...)
pub fn advance(self: *Parser) ParseError!?TokInfo {
    var res: ?TokInfo = null;
    if (self.toks.items.len > 0) {
        res = self.toks.orderedRemove(0);
    }
    return res;
}

pub fn peek(self: *Parser, matchable: Matchable) !bool {
    if (self.ended) {
        return false;
    }
    return try self.nth(0) != null and matchable.eql(try self.nth(0).?.tok, self);
}

pub fn peekAny(self: *Parser, matchables: []const Matchable) bool {
    for (matchables) |matchable| {
        if (self.peek(matchable)) return true;
    }
    return false;
}

/// Are these exact options going to happen in order?
pub fn peekAll(self: *Parser, matchables: []const Matchable) bool {
    for (matchables) |matchable, i| {
        if (try self.nth(i) == null or !matchable.eql(try self.nth(i).?.tok, self)) return false;
    }
    return true;
}

pub fn peekBlockStartingWith(self: *Parser, matchable: Matchable) bool {
    return self.peekAll(&[_]Matchable{
        .{ .token = .{ .newline = .{} } },
        .{ .token = .{ .indent = .{} } },
        .{ .any_access = .{} },
    });
}

pub fn tryMatch(self: *Parser, matchable: Matchable) ParseError!?Token {
    if (self.peek(matchable)) return self.advance();
    return null;
}

pub fn tryMatchAny(self: *Parser, matchables: []const Matchable) ParseError!?Token {
    if (self.peekAny(matchables)) return try self.advance();
    return null;
}

pub fn tryMatchAll(self: *Parser, matchables: []const Matchable) !bool {
    if (self.peekAll(matchables)) {
        try self.advanceN(matchables.len);
        return true;
    } else return false;
}

pub fn tryMatchBlockStartingWith(self: *Parser, matchable: Matchable) ParseError!?Token {
    if (self.peekBlockStartingWith(matchable)) {
        try self.advanceN(2);
        return try self.advance();
    } else return false;
}

pub fn match(self: *Parser, matchable: Matchable) ParseError!Token {
    const res = try self.tryMatch(matchable);
    if (res) |r| return r;
    return error.Unmatched;
}

pub fn matchAny(self: *Parser, matchables: []const Matchable) ParseError!Token {
    if (try self.tryMatchAny(matchables)) |m| return m;
    self.cache.dump();
    printf("\nExpected any of {any}, found {any}\n\n", .{ matchables, try self.nth(0) });
    return error.Unmatched;
}

pub fn matchAll(self: *Parser, matchables: []const Matchable) ParseError!void {
    if (!try self.tryMatchAll(matchables)) return error.Unmatched;
}

pub fn matchBlockStartingWith(self: *Parser, matchable: Matchable) ParseError!Token {
    if (try self.tryMatchBlockStartingWith(matchable)) |m| return m;
    return error.Unmatched;
}

pub fn parseExpr(self: *Parser) ParseError!Atom {
    return self.parseBinary(0);
}

pub fn parseBinary(self: *Parser, prec: usize) ParseError!Atom {
    var res = try self.descend(prec);
    while (try self.tryMatch(.{ .prec = .{ .binary = prec } })) |op| {
        res = try common.cmdID(op.id().?, &.{ res, try self.descend(prec) });
    }

    return res;
}

pub fn parseUnary(self: *Parser) ParseError!Atom {
    // printf("\nParsing unary\n", .{});
    var res = try self.parseHookable();
    const should_unary = all(&.{
        try self.nth(0) != null,
        !self.ended,
        !self.peekAny(STOPPERS[0..]),
        !self.peek(.{ .any_binary = .{} }),
        !self.peek(.{ .any_assg = .{} }),
    });
    if (should_unary) {
        // printf("\nParsing unary argument...\n", .{});
        res = try common.list(res, &.{try self.parseUnary()});
    }
    return res;
}

/// This is where the magic hooking for stuff like `if` happens.
pub fn parseHookable(self: *Parser) ParseError!Atom {
    var res: ?Atom = null;
    if (self.cur[0] != null) {
        if (self.hooks.get(self.cur[0].?.tok)) |hook| {
            res = try hook(self);
        }
    }
    if (res == null) res = try self.parseParticle();

    try self.parseTrailers(&res.?);
    return res.?;
}

// Valid example:
//from(foos)
//  .select: foo => foo.x
//  .where: x => x > 10
//  .order_by: x => x, #Desc

// TODO: WIP
// Parse any series of calls and accesses. At the very end can be either an access block or a `:` call.
pub fn parseTrailers(self: *Parser, res: *Atom) ParseError!void {
    const any_access = Matchable{ .any_access = .{} };
    var clean = false;
    while (!clean) {
        clean = true;
        if (self.peek(.{ .any_access = .{} })) {
            clean = false;
        }
        if (self.peekAny(&CALLERS) and self.cur[0].?.attached) {
            clean = false;
        }
    }

    if (try self.tryMatch(.{ .common = .{ .punc = .iColon } }) != null) {
        if (try self.tryMatch(.{ .seq_of = &.{
            .{ .token = .{ .newline = .{} } },
            .{ .token = .{ .indent = .{} } },
        } }) != null) {
            // TODO: Make this properly append the block as an argument if this was appended to a call
            res.* = try common.list(res.*, &.{try self.parseStmtSeq()});
            _ = try self.match(.{ .seq_of = &.{
                .{ .token = .{ .newline = .{} } },
                .{ .token = .{ .dedent = .{} } },
            } });
        } else res.* = try common.list(res.*, &.{try self.parseExpr()});
        self.ended = true;
    } else if (self.peekBlockStartingWith(any_access)) {
        try self.advanceN(2); // enter the block

        // Each line starts with an access
        while (true) {
            const accessor = try self.match(any_access);
            res.* = try common.cmdID(accessor.id().?, &.{ res.*, try self.parseParticle() });
            // Just recurse to deal with all the subsequent calls/accesses on that line
            try self.parseTrailers(res);

            if (self.peek(.{ .token = .dedent })) break;
        }

        _ = try self.match(.{ .token = .{ .dedent = .{} } });
        self.ended = true;
    }
}

/// Word, tag, tuple, block, or array
pub fn parseParticle(self: *Parser) ParseError!Atom {
    const res = try self.matchAny(&PARTICLES);
    switch (res) {
        .word => |id| return common.word(id),
        // .tag => |id| return Atom{ .tag = id },
        else => unreachable,
    }
}

/// (@block as stmt1 stmt2...)
pub fn parseStmtSeq(self: *Parser) ParseError!Atom {
    var res = try common.cmdCommon(.ibBlock, &.{});

    while (true) {
        printf("Top of seq loop\n", .{});
        while ((try self.tryMatch(.{ .token = .newline })) != null or (try self.tryMatch(.{ .common = .{ .punc = .iSemicolon } })) != null) {}

        // printf("Begin new stmtSeqPart\n", .{});
        try res.list.append(try self.parseExpr());

        if (self.ended) {
            self.ended = false;
        }

        if ((try self.tryMatch(.{ .token = .newline })) == null and (try self.tryMatch(.{ .common = .{ .punc = .iSemicolon } })) == null) {
            // printf("Found a {}, and so did not continue in the seq.\n", .{self.cur[0]});
            break;
        }

        if (self.cur[0] == null) {
            break;
        }
    }

    return res;
}

pub fn expected(self: *Parser, what: Matchable) ParseError!void {
    printf("\nError at {}: Expected ", .{self.cur[0].?.pos});
    switch (what) {
        .token => |tok| printf("{}", .{tok}),
        .common => |com| printf("{}", .{com.toStr()}),
        .prec => |com| {
            switch (com) {
                .access => printf("an access operator", .{}),
                .binary => |p| printf("a binary operator of precedence {}", .{p}),
            }
        },
        .any_binary => printf("any binary operator", .{}),
        .any_access => printf("any access operator", .{}),
        .any_word => printf("any word", .{}),
    }
    printf(", but found {}\n\n", .{self.cur[0].?.tok});
    return error.Expected;
}

// As a workaround for zigfmt's zealousness around samelining everything.
pub fn all(conds: []const bool) bool {
    for (conds) |cond| {
        if (!cond) return false;
    }
    return true;
}

pub fn precOf(self: Parser, tok: Token, attached: bool) Prec {
    if (self.precs.tryGetValue(tok)) |prec| return prec;

    return .{ .none = .{} };
}

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║Internals                                                                  ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
fn descend(self: *Parser, prec: usize) ParseError!Atom {
    if (prec == self.precs.count()) {
        return self.parseUnary();
    } else {
        return self.parseBinary(prec + 1);
    }
}

// Things which can unambiguously stop the current binary/unary expression.
// `:`, `)`, `}`, `]`, `,`, `;`, and newline
pub const STOPPERS = [_]Matchable{
    .{ .common = .{ .word = .iColon } },
    .{ .common = .{ .word = .iRParen } },
    .{ .common = .{ .word = .iRBrace } },
    .{ .common = .{ .word = .iRBracket } },
    .{ .common = .{ .word = .iComma } },
    .{ .common = .{ .word = .iSemicolon } },
    .{ .token = .{ .newline = .{} } },
};

pub const ENDERS = [_]Matchable{
    .{ .token = .{ .newline = .{} } },
    .{ .common = .{ .word = .iSemicolon } },
};
pub const CALLERS = [_]Matchable{
    .{ .common = .{ .word = .iLParen } },
    .{ .common = .{ .word = .iLBracket } },
};

pub const PARTICLES = [_]Matchable{
    .{ .any_word = .{} },
};

pub const COMMON_PRECS = comptime [_][]const CPrec{
    &[_]CPrec{
        // Assg and Access
        .{ .tok = .{ .common = .iAccess }, .prec = .access },
        .{ .tok = .{ .common = .iAssg }, .prec = .assg },
        .{ .tok = .{ .common = .iSubAssg }, .prec = .assg },
        .{ .tok = .{ .common = .iAddAssg }, .prec = .assg },
        .{ .tok = .{ .common = .iMulAssg }, .prec = .assg },
        .{ .tok = .{ .common = .iDivAssg }, .prec = .assg },
    },

    // All binaries grouped by precedence
    &[_]CPrec{
        .{ .tok = .{ .common = .iLambda } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iOr } },
        .{ .tok = .{ .common = .iXor } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iAnd } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iEq } },
        .{ .tok = .{ .common = .iNeq } },
        .{ .tok = .{ .common = .iGtEq } },
        .{ .tok = .{ .common = .iLtEq } },
        .{ .tok = .{ .common = .iLt } },
        .{ .tok = .{ .common = .iGt } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iIn } },
        .{ .tok = .{ .common = .iNotIn } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iOpenRange } },
        .{ .tok = .{ .common = .iClosedRange } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iAdd } },
        .{ .tok = .{ .common = .iSub } },
    },
    &[_]CPrec{
        .{ .tok = .{ .common = .iMul } },
        .{ .tok = .{ .common = .iDiv } },
    },
};

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║Tests                                                                      ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
test "basic parsing" {
    printf("\n\n", .{});
    try common.initCommons();
    var puncs: [Lexer.COMMON_CPUNCS.len]Lexer.Punc = undefined;
    for (Lexer.COMMON_CPUNCS) |cpunc, i|
        puncs[i] = try cpunc.resolve(&common.cache);
    var lexer = Lexer.init(
        @embedFile("../tests/parsing.li"),
        &puncs,
        &common.cache,
        common.alloc,
    );
    var parser = Parser.init(&lexer, common.alloc);
    for (COMMON_PRECS[0]) |cprec| {
        if (cprec.prec == .assg) {
            try parser.precs.put(try cprec.tok.resolve(&common.cache), .{ .assg = .{} });
        } else if (cprec.prec == .access) {
            try parser.precs.put(try cprec.tok.resolve(&common.cache), .{ .access = .{} });
        }
    }
    for (COMMON_PRECS[1..]) |cprecs, i| {
        for (cprecs) |cprec|
            try parser.precs.put(try cprec.tok.resolve(&common.cache), .{ .binary = i });
    }

    _ = try parser.advance();
    _ = try parser.advance();
    _ = try parser.advance();
    common.cache.dump();
    const root = try parser.parseStmtSeq();
    printf("Root: \n", .{});
    root.print(&common.cache, 0);
    printf("\n\n", .{});
}
